File: ClassEval_T/cpp/solution/CurrencyConverter.cpp
#include <unordered_map>
#include <string>
#include <vector>

class CurrencyConverter {
public:
    CurrencyConverter();

    double convert(double amount, const std::string& from_currency, const std::string& to_currency);
    std::vector<std::string> get_supported_currencies();
    bool add_currency_rate(const std::string& currency, double rate);
    bool update_currency_rate(const std::string& currency, double new_rate);

    std::unordered_map<std::string, double> rates;
    std::vector<std::string> currency_order;
};

CurrencyConverter::CurrencyConverter() {
    rates = {
        {"USD", 1.0},
        {"EUR", 0.85},
        {"GBP", 0.72},
        {"JPY", 110.15},
        {"CAD", 1.23},
        {"AUD", 1.34},
        {"CNY", 6.40},
    };

    currency_order = { "USD", "EUR", "GBP", "JPY", "CAD", "AUD", "CNY" };
}

double CurrencyConverter::convert(double amount, const std::string& from_currency, const std::string& to_currency) {
    if (from_currency == to_currency) {
        return amount;
    }

    if (rates.find(from_currency) == rates.end() || rates.find(to_currency) == rates.end()) {
        return false;
    }

    double from_rate = rates[from_currency];
    double to_rate = rates[to_currency];

    double converted_amount = (amount / from_rate) * to_rate;
    return converted_amount;
}

std::vector<std::string> CurrencyConverter::get_supported_currencies() {
    return currency_order;
}

bool CurrencyConverter::add_currency_rate(const std::string& currency, double rate) {
    if (rates.find(currency) != rates.end()) {
        return false;
    }
    rates[currency] = rate;
    currency_order.push_back(currency);
    return true;
}

bool CurrencyConverter::update_currency_rate(const std::string& currency, double new_rate) {
    if (rates.find(currency) == rates.end()) {
        return false;
    }
    rates[currency] = new_rate;
    return true;
}

File: ClassEval_T/java/solutuon/CurrencyConverter.java
package org.example;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class CurrencyConverter {
    private Map<String, Double> rates;

    public CurrencyConverter() {
        rates = new HashMap<>();
        rates.put("USD", 1.0);
        rates.put("EUR", 0.85);
        rates.put("GBP", 0.72);
        rates.put("JPY", 110.15);
        rates.put("CAD", 1.23);
        rates.put("AUD", 1.34);
        rates.put("CNY", 6.40);
    }

    public double convert(double amount, String fromCurrency, String toCurrency) {
        if (fromCurrency.equals(toCurrency)) {
            return amount;
        }

        if (!rates.containsKey(fromCurrency) || !rates.containsKey(toCurrency)) {
            return -1;
        }

        double fromRate = rates.get(fromCurrency);
        double toRate = rates.get(toCurrency);

        double convertedAmount = (amount / fromRate) * toRate;
        return convertedAmount;
    }

    public Set<String> getSupportedCurrencies() {
        return rates.keySet();
    }

    public boolean addCurrencyRate(String currency, double rate) {
        if (rates.containsKey(currency)) {
            return false;
        }
        rates.put(currency, rate);
        return true;
    }

    public boolean updateCurrencyRate(String currency, double newRate) {
        if (!rates.containsKey(currency)) {
            return false;
        }
        rates.put(currency, newRate);
        return true;
    }

    public Map<String, Double> getRates() {
        return rates;
    }
}